File: src/engine/hooks.ts
// engine/hooks.ts — イベントフックのディスパッチ
// 戦闘内イベントに反応するフック (リスナー) をここで束ねる。
// 現在のフック: ①置物 (permanent) ②リアクション方式。
// レリック・敵パッシブなどは将来ここに追加する (CLAUDE.md「イベントフックの口だけは最初から開けておく」)。

import { runPermanentTriggers } from './effects.ts'
import { getReactionSystem } from './reactions/index.ts'
import type { GameEvent, GameState } from './types.ts'

// 置物トリガーの実行本体は effects.ts へ移設 (回復・HP損失・消滅の誘発を効果解決の内側から発火させるため)。
// combat.ts などの既存の参照は従来どおりここから import できる
export { runPermanentTriggers }

export function dispatchHooks(state: GameState, event: GameEvent): GameState {
  let s = state
  // 1. 置物フック (自動発火。リアクションの割り込みより先に解決する)
  if (event.type === 'EnemyActionExecuting' && event.kind === 'attack') {
    // 被攻撃前 (軽減系) の置物
    s = runPermanentTriggers(s, 'onAttackIncoming', event.enemyIndex)
  }
  if (event.type === 'EnemyActionResolved' && event.kind === 'attack') {
    // 被攻撃後 (返し系) の置物: 茨の茂みなど
    s = runPermanentTriggers(s, 'onAttacked', event.enemyIndex)
  }
  if (event.type === 'EnemyActionExecuting' || event.type === 'EnemyActionResolved') {
    // 置物の返しで敵が倒れた・自分が倒れた場合、リアクション確認はもう不要
    const enemy = s.enemies[event.enemyIndex]
    if ((enemy && enemy.hp <= 0) || s.player.hp <= 0) return s
  }
  // 2. リアクション方式フック (方式は state.reactionMode から解決)
  s = getReactionSystem(s.reactionMode).onEvent(s, event)
  // 3. (将来) レリックフック / 敵パッシブフック をここに追加
  return s
}

File: unity/Packages/com.deckrogue.engine/Runtime/Hooks.cs
// Hooks.cs — src/engine/hooks.ts の手書き移植 (unity/PORTING.md)
// 戦闘内イベントに反応するフック (リスナー) をここで束ねる。
// 現在のフック: ①置物 (permanent) ②リアクション方式。
using DeckRogue.Engine.Generated;

namespace DeckRogue.Engine
{
    public static class Hooks
    {
        /// <summary>置物トリガーの実行本体は Effects へ移設。既存の参照は従来どおりここから呼べる</summary>
        public static GameState RunPermanentTriggers(GameState state, string trigger, int enemyIndex)
            => Effects.RunPermanentTriggers(state, trigger, enemyIndex);

        public static GameState DispatchHooks(GameState state, GameEvent ev)
        {
            var s = state;
            // 1. 置物フック (自動発火。リアクションの割り込みより先に解決する)
            if (ev is GameEvent_EnemyActionExecuting exec1 && exec1.Kind == EnemyActionKinds.Attack)
            {
                // 被攻撃前 (軽減系) の置物
                s = Effects.RunPermanentTriggers(s, "onAttackIncoming", exec1.EnemyIndex);
            }
            if (ev is GameEvent_EnemyActionResolved res1 && res1.Kind == EnemyActionKinds.Attack)
            {
                // 被攻撃後 (返し系) の置物: 茨の茂みなど
                s = Effects.RunPermanentTriggers(s, "onAttacked", res1.EnemyIndex);
            }
            if (ev is GameEvent_EnemyActionExecuting || ev is GameEvent_EnemyActionResolved)
            {
                // 置物の返しで敵が倒れたらリアクション確認はもう不要。
                // プレイヤーのHPが0以下でも post窓は開く (致死時の誘発)
                int idx = ev is GameEvent_EnemyActionExecuting e2 ? e2.EnemyIndex : ((GameEvent_EnemyActionResolved)ev).EnemyIndex;
                var enemy = (idx >= 0 && idx < s.Enemies.Count) ? s.Enemies[idx] : null;
                if (enemy != null && enemy.Hp <= 0) return s;
                if (s.Player.Hp <= 0 && ev is GameEvent_EnemyActionExecuting) return s;
            }
            // 2. リアクション方式フック (方式は state.reactionMode から解決)
            s = ReactionSystems.Get(s.ReactionMode).OnEvent(s, ev);
            // 3. (将来) レリックフック / 敵パッシブフック をここに追加
            return s;
        }
    }
}
